### PR TITLE
#34: Admin Client login screen

### LINKED ISSUE #34: Login Screen
User Stories to satisfy:
------------

Given that I am on the admin client
And it is my first time accessing the system
I want to be shown the login screen

Given that I am on the admin client
And I am not logged in
I want to be shown the login screen

Given that I am on the admin client
And I have logged in but my token exipred a while ago
I want to be shown the login screen

Given that I am on the admin client
And I have a valid token
I want to be shown the main screen instead of the login screen

Given that I am on the login screen
And I log in with correct credentials
And I have the "Admin Client" permissions
I want to be shown the main screen instead of the login screen

Given that I am on the login screen
And I log in with correct credentials
And I do not have the "Admin Client" permissions
I want to be shown an error message saying that I do not have the necessary permissions to access the application

Given that I am on the login screen
And I log in with incorrect credentials
I want to be shown an error message saying that my credentials are incorrect

Details:
---------

The first screen a user must see is a login screen. This will accept a username and password, and obtain a token from the backend to use for subsequent requests to the backend. This token will also specify what the user can and cannot see.

When this token does not exist or it is expired, all pages must redirect to the login page.

### PR DESCRIPTION
**Pull request by markscamilleri**
Resolves #34 

### Style Rules
   - Turned off 2 rules which were conflicting with another rule

### Dependencies
  - Added `@angular/flex-layout`
  - Added `@auth0/angular-jwt` 
  - Removed `@ionic/angular`

### Page
  - Auth Page and routing, with tests

### RepClient
 - CodeFactor automatically fixed some style issues - blame it not me!

## Screenshots

Auth Screen - Snack bar at the bottom when one tries to access a protected page without being authenticated
Auth Screen - Snack bar at the bottom when one tries to access a protected page without being authenticated

Auth screen - After snack bar disappears (or if accessed directly)
Auth screen - After snack bar disappears (or if accessed directly)

Auth screen - Log in button being activated when input
Auth screen - Log in button being activated when input

Auth screen - Logging in with loading overlay
Auth screen - Logging in with loading overlay

### CODE REVIEW

**Comment by markscamilleri** [comment]
Seems like Travis is under maintenance.

Also, I wasn't sure how to test that once the form submits the user is redirected to the page they wanted to go to - everytime I tested it with Jasmine and Karma, the test behaved unexpectedly (at one point even triggering Karma to generate infinite test cases). However, I did test it manually and it works.

So if a user is redirected to `/auth` from, say, `/customers` because they're not logged in, they'll be redirected back once they logged in. Permissions need to be checked (see #72 )

**Comment by markscamilleri** [comment]
Regarding the screenshots: I just realized there's a square income with 3 dots. That's a browser extension of mine and has nothing to do with the page. I can update the screenshots if you prefer

File: AdminClient/src/app/auth/auth.component.ts [hunk 1]
```diff
@@ -75,31 +93,25 @@ export class AuthPage implements OnInit, OnDestroy {
...
+    this.loadingSpinnerService.spin$.next(false);
   }
 
   protected isFormValid(): boolean {
     return this.authForm.valid;
   }
 
   /**
-   * Returns appropriate error message for email validation
+   * @returns string with the appropriate error message for email validation
```
Comment (HarryPi) [inline AdminClient/src/app/auth/auth.component.ts:114]: @returns is redundant with TypeScript

File: AdminClient/src/app/app.module.ts [hunk 2]
```diff
@@ -1,37 +1,63 @@
...
     DashboardComponent,
     CustomersComponent
   ],
+  bootstrap: [AppComponent],
   imports: [
+    AppRoutingModule,
+    AuthModule,
     BrowserModule,
+    BrowserAnimationsModule,
+    EffectsModule.forRoot([AuthEffects]),
```
Comment (HarryPi) [inline AdminClient/src/app/app.module.ts:41]: Are you guys using NgRx store? If not this is redundant
```suggestion
    
```

File: AdminClient/src/app/app.module.ts [hunk 3]
```diff
@@ -1,37 +1,63 @@
...
-    BrowserAnimationsModule,
-    MatButtonModule,
-    MatIconModule
+    StoreModule.forRoot(appReducer, {
+      metaReducers,
+      runtimeChecks: {
+        strictActionImmutability: true,
+        strictStateImmutability: true
+      }
+    })
```
Comment (HarryPi) [inline AdminClient/src/app/app.module.ts:59]:
```suggestion
    
```
Again redundant if no NgRx

File: AdminClient/src/app/auth/auth.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    mockStore.setState({
+      auth: {
+        errorMessage: null,
+        loading: true,
+        user: null,
+        returnUrl: '',
+      }
+    });
+
+    mockStore.refreshState();
```
Comment (HarryPi) [inline AdminClient/src/app/auth/auth.component.spec.ts:92]:
```suggestion

```

NgRx Code

File: AdminClient/src/app/auth/auth.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    mockStore.setState({
+      auth: {
+        errorMessage: 'Some Error message',
+        loading: false,
+        user: null,
+        returnUrl: '',
+      }
+    });
+
+    mockStore.refreshState();
```
Comment (HarryPi) [inline AdminClient/src/app/auth/auth.component.spec.ts:70]:
```suggestion
    
```

NgRx code

File: AdminClient/src/app/auth/auth.component.ts [hunk 6]
```diff
@@ -25,16 +26,22 @@ export class AuthPage implements OnInit, OnDestroy {
...
+  /**
+   * The url to go to once a login is successful
+   */
+  private returnUrl: string;
+
   constructor(
     public loadingSpinnerService: LoadingSpinnerService,
+    private route: ActivatedRoute,
+    private router: Router,
     private store: Store<AppState>,
```
Comment (HarryPi) [inline AdminClient/src/app/auth/auth.component.ts:38]:
```suggestion
  
```

**Comment by markscamilleri** [comment]
> I am a bit confused are you guys planning on using NgRx store for the entire application, i cant really review this without understanding this as it is confusing.

Yes. We're using it to trigger the loading screen and to fire events - the Auth uses this to redirect the user back to where they wanted to go on a successful login. Furthermore, errors saying the user is not logged in and then forwarded to Auth are handled with ngrx.

> Also a small detail @returns in jsdoc is redundant with TypeScript as there are types

It would be if I'd use it to document the types. In this case, it's for documentation to explain what the return value means (and such docs could be generated if need be - helps if and when someone works on the page after me)

**Comment by HarryPi** [comment]
> Seems like Travis is under maintenance.
> 
> Also, I wasn't sure how to test that once the form submits the user is redirected to the page they wanted to go to - every time I tested it with Jasmine and Karma, the test behaved unexpectedly (at one point even triggering Karma to generate infinite test cases). However, I did test it manually and it works.
> 
> So if a user is redirected to `/auth` from, say, `/customers` because they're not logged in, they'll be redirected back once they logged in. Permissions need to be checked ( #72 )

Regarding this this can be done by using the mock angular router.
Basically you spy on the redirect of the router and check if the redirect function was called.

**Comment by HarryPi** [comment]
> > I am a bit confused are you guys planning on using NgRx store for the entire application, i cant really review this without understanding this as it is confusing.
> 
> Yes. We're using it to trigger the loading screen and to fire events - the Auth uses this to redirect the user back to where they wanted to go on a successful login. Furthermore, errors saying the user is not logged in and then forwarded to Auth are handled with ngrx.

So are you guys going to use it for anything else in the App ? if not isnt it a bit of an overkill this could be simply converted to a service. NgRx store is not a lighweight package

**Comment by HarryPi** [comment]
> Regarding the screenshots: I just realized there's a square income with 3 dots. That's a browser extension of mine and has nothing to do with the page. I can update the screenshots if you prefer

That is fine

**Comment by markscamilleri** [comment]
> > Seems like Travis is under maintenance.
> > 
> > Also, I wasn't sure how to test that once the form submits the user is redirected to the page they wanted to go to - every time I tested it with Jasmine and Karma, the test behaved unexpectedly (at one point even triggering Karma to generate infinite test cases). However, I did test it manually and it works.
> > 
> > So if a user is redirected to `/auth` from, say, `/customers` because they're not logged in, they'll be redirected back once they logged in. Permissions need to be checked ( #72 )
> 
> Regarding this this can be done by using the mock angular router.
> Basically you spy on the redirect of the router and check if the redirect function was called.
> 
> 

I'll check it out. I tried to but I was unsuccessful as Karma kept restarting whenever the form was submitted and the user redirected.

**Comment by markscamilleri** [comment]
> > > I am a bit confused are you guys planning on using NgRx store for the entire application, i cant really review this without understanding this as it is confusing.
> > 
> > 
> > Yes. We're using it to trigger the loading screen and to fire events - the Auth uses this to redirect the user back to where they wanted to go on a successful login. Furthermore, errors saying the user is not logged in and then forwarded to Auth are handled with ngrx.
> 
> So are you guys going to use it for anything else in the App ? if not isnt it a bit of an overkill this could be simply converted to a service. NgRx store is not a lighweight package

No. We are triggering the loading srcreen through state with NgRx, idk if we can do it another way? maybe just an internal state for the service or just triggering the function?

With regards to the Errors and the redirect - I know we can do the redirect in a different way without ngRx, but I'm not sure about how to do the snackbar error?

**Comment by markscamilleri** [comment]
Re-review is not needed, had to merge master which only had a Readme File and some attachments for the wiki.

**Comment by HarryPi** [comment]
> Re-review is not needed, had to merge master which only had a Readme File and some attachments for the wiki.

Yes this is perfectly fine

**Pull request merged**